### PR TITLE
Bootstrapping Php client codegen

### PR DESCRIPTION
**Pull request by garrettjonesgoogle**
@shinfan

### CODE REVIEW

File: vgen/src/main/java/io/gapi/vgen/php/PhpContext.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gapi.vgen.php;
+
+/**
+ * A PhpContext provides functionality specific to a context in PHP.
+ */
+public interface PhpContext {
+  void resetState(PhpSnippetSet<?> phpSnippetSet, PhpContextCommon phpCommon);
```
Comment (shinfan) [inline vgen/src/main/java/io/gapi/vgen/php/PhpContext.java]: Some comments here will be helpful

File: vgen/src/main/java/io/gapi/vgen/php/PhpGapicContext.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+  }
+
+  public String getServiceName(Interface service) {
+    return service.getFullName();
+  }
+
+  /**
+   * Gets the full name of the message or enum type in Java.
+   */
+  public String getTypeName(ProtoElement elem) {
```
Comment (shinfan) [inline vgen/src/main/java/io/gapi/vgen/php/PhpGapicContext.java]: This method is not used.

File: vgen/src/main/java/io/gapi/vgen/php/PhpGapicContext.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  // ===============
+
+  /**
+   * Returns the PHP filename which holds the gRPC service definition.
+   */
+  public String getGrpcFilename(Interface service) {
+    return service.getFile().getProto().getName().replace(".proto", "_services");
+  }
+
+  public String getTypeName(String typeName) {
```
Comment (shinfan) [inline vgen/src/main/java/io/gapi/vgen/php/PhpGapicContext.java]: Can this method be private since it is not used from the outside?

Comment (garrettjonesgoogle) [inline vgen/src/main/java/io/gapi/vgen/php/PhpGapicContext.java]: It will likely be used in the snippet file later when it is fleshed out more.

File: vgen/src/main/java/io/gapi/vgen/php/PhpContext.java (hunk 1 above)
Comment (garrettjonesgoogle) [inline vgen/src/main/java/io/gapi/vgen/php/PhpContext.java]: The equivalent class in Java codegen (JavaContext) doesn't have this commented either. I'd prefer to make a documentation pass in a separate PR.

File: vgen/src/main/java/io/gapi/vgen/php/PhpContextCommon.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.List;
+
+public class PhpContextCommon {
+
+  /**
+   * A bi-map from full names to short names indicating the import map.
+   */
+  private final BiMap<String, String> imports = HashBiMap.create();
+
+  public String getMinimallyQualifiedName(String fullName, String shortName) {
```
Comment (shinfan) [inline vgen/src/main/java/io/gapi/vgen/php/PhpContextCommon.java]: This method does two things:
(1) find the minimally qualified name -> I can tell this from the method name
(2) add the full name -> short name to the imports map -> I cannot tell this from the method name since it has the prefix 'get'

I suggest to have a separate method to handle (2)

File: vgen/src/test/java/io/gapi/vgen/PhpCodeGeneratorTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.util.List;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+
+/**
+ * Ruby code generator baseline tests.
```
Comment (shinfan) [inline vgen/src/test/java/io/gapi/vgen/PhpCodeGeneratorTest.java]: Ruby / PHP code generator

File: vgen/src/main/java/io/gapi/vgen/php/PhpContextCommon.java (hunk 4 above)
Comment (garrettjonesgoogle) [inline vgen/src/main/java/io/gapi/vgen/php/PhpContextCommon.java]: There is an equivalent method in JavaContextCommon. I'd prefer to consider a change in approach separate from this PR and apply it to both languages.

File: vgen/src/main/java/io/gapi/vgen/php/PhpGapicContext.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    // TODO get the title from the service yaml file
+    return "";
+  }
+
+  public String getServiceName(Interface service) {
+    return service.getFullName();
+  }
+
+  /**
+   * Gets the full name of the message or enum type in Java.
```
Comment (shinfan) [inline vgen/src/main/java/io/gapi/vgen/php/PhpGapicContext.java]: Java / PHP ?

File: vgen/src/main/java/io/gapi/vgen/php/PhpGapicContext.java (hunk 2 above)
Comment (garrettjonesgoogle) [inline vgen/src/main/java/io/gapi/vgen/php/PhpGapicContext.java]: I will remove it for now.

**Comment by garrettjonesgoogle** [comment]
PTAL

**Comment by shinfan** [comment]
LGTM

**Comment by garrettjonesgoogle** [comment]
Filed: https://github.com/googleapis/toolkit/issues/62

**Comment by garrettjonesgoogle** [comment]
Also filed: https://github.com/googleapis/toolkit/issues/63
